File: engine-python/src/ibkr_agent/alerts.py
"""价位警告 —— 纯计算 + 状态机,离线可单测。

给一个标的挑出**值得盯的价位**,再在价格穿过它们时报一次。价位有三个来源:

  期权墙   当天期权链上持仓/成交最密的行权价、最大痛点、gamma 翻转位。
           这些位置背后有真实的对冲需求,不是画出来的。
  趋势位   日线均线(60/120/200 日)与 52 周高低点。跌到年线、创 52 周新低,
           都是全市场共用的"历史低位"标尺——大量资金把它们当决策触发点,
           所以价格到这儿真的会有反应。
  整数关口 现价上下最近的 5 的整数倍(可调)。这类位置没有基本面理由,
           但挂单确实爱堆在整数上——IREN 41 块,40 就是这么一个位置。

**这个模块最难的不是"设警报",是"不要刷屏"。** 价格在 40 附近来回蹭一分钟,
朴素实现能报几十次,报到第三次人就不看了,警告也就废了。所以每个价位有自己的
状态机:穿过 → 报一次 → **落防(disarmed)** → 价格离开足够远才重新上膛。
再叠一层冷却时间兜底。

只读研究模块:只发通知,不下单、不改仓位。
"""
from __future__ import annotations

from dataclasses import dataclass, field
from math import ceil, floor
from typing import Any, Dict, List, Optional, Sequence, Tuple

DEFAULT_STEP = 5.0        # 整数关口的步长(用户可改)
DEFAULT_MERGE_PCT = 0.15  # 两个价位相距小于现价的这个百分比就并成一个
DEFAULT_BAND_PCT = 0.30   # 落防后要离开价位这么远(占价位的百分比)才重新上膛
DEFAULT_COOLDOWN = 300.0  # 同一个价位两次报警的最小间隔(秒),兜底用
MAX_LEVELS = 24           # 一个标的最多盯这么多价位,再多就是噪音

DEFAULT_MA_PERIODS = (20, 60, 120, 200)  # 日线均线:20/60 是用户天天盯的,120 是季线,200 是全市场盯的
EXTREME_WINDOW = 250                 # 52 周 ≈ 250 个交易日
MIN_EXTREME_BARS = 20                # 历史太短时高低点没意义,干脆不给


class AlertError(ValueError):
    pass


@dataclass(frozen=True)
class AlertLevel:
    price: float
    label: str
    source: str      # call_wall / put_wall / call_vol_wall / put_vol_wall
                     # / max_pain / gamma_flip / round
    kind: str        # resistance / support / pivot
    priority: int = 0   # 合并时谁说了算,越大越可信(见 _wall_levels)

    def as_dict(self) -> Dict[str, Any]:
        return {"price": self.price, "label": self.label,
                "source": self.source, "kind": self.kind}


@dataclass
class LevelState:
    """一个价位的上膛状态。armed=False 表示刚报过,还没离开得够远。"""

    armed: bool = True
    last_fired_at: Optional[float] = None

    def as_dict(self) -> Dict[str, Any]:
        return {"armed": self.armed, "last_fired_at": self.last_fired_at}


# ---------------------------------------------------------------- 价位生成
def round_levels(spot: float, step: float = DEFAULT_STEP) -> List[float]:
    """现价上下最近的整数关口。

    现价正好落在关口上时,那个关口本身没有意义(警报会立刻触发),
    所以改取它上下各一档。
    """
    if spot <= 0 or step <= 0:
        return []
    below = floor(spot / step) * step
    above = ceil(spot / step) * step
    if abs(above - below) < 1e-9:                 # 现价正好在关口上
        return [round(below - step, 6), round(above + step, 6)]
    return [round(below, 6), round(above, 6)]


# 合并优先级。数字只在两个价位靠得太近、要并成一个时起作用:留谁的价、
# 谁的标签排前面。整数关口永远垫底——它没有任何持仓依据。
# 趋势位(均线/52周高低点)与弱墙同级:都是全市场看得见、但不带当天持仓
# 依据的位置。同级贴在一起时按价排序后靠前的赢(合并逻辑保留 previous)。
_PRIORITY_ROUND = 0
_PRIORITY_PIVOT = 1
_PRIORITY_WEAK_WALL = 2
_PRIORITY_TREND = 2
_PRIORITY_STRONG_WALL = 3


def _wall_levels(wall: Optional[Dict[str, Any]]) -> List[AlertLevel]:
    """把期权墙分析折成价位。

    0DTE 时成交量墙比持仓墙可信(OI 是隔夜存量,反映不了当天的流),
    其余到期日反过来。这个差别落在**合并优先级**上:两种墙贴在一起时,
    以更可信的那个的价位为准。
    """
    if not wall:
        return []
    out: List[AlertLevel] = []
    zero_day = (wall.get("days_to_expiry") or 99) <= 1.0
    vol_rank = _PRIORITY_STRONG_WALL if zero_day else _PRIORITY_WEAK_WALL
    oi_rank = _PRIORITY_WEAK_WALL if zero_day else _PRIORITY_STRONG_WALL

    def add(key: str, label: str, rank: int) -> None:
        entry = wall.get(key)
        if entry and entry.get("strike"):
            kind = "resistance" if "call" in key else "support"
            out.append(AlertLevel(float(entry["strike"]), label, key, kind, rank))

    add("call_vol_wall", "上方成交墙", vol_rank)
    add("put_vol_wall", "下方成交墙", vol_rank)
    add("call_wall", "上方持仓墙", oi_rank)
    add("put_wall", "下方持仓墙", oi_rank)

    if wall.get("max_pain"):
        out.append(AlertLevel(float(wall["max_pain"]["strike"]), "最大痛点",
                              "max_pain", "pivot", _PRIORITY_PIVOT))
    if wall.get("gamma_flip") is not None:
        out.append(AlertLevel(float(wall["gamma_flip"]), "Gamma 翻转位",
                              "gamma_flip", "pivot", _PRIORITY_PIVOT))
    return out


def _trend_kind(price: float, spot: float) -> str:
    """价位在现价下方是支撑,上方是阻力,正好贴着算 pivot。"""
    if price < spot:
        return "support"
    if price > spot:
        return "resistance"
    return "pivot"


def _bar_values(bars: Optional[Sequence[Dict[str, Any]]], key: str) -> List[float]:
    """按顺序取一列有效价格。缺了或非正数的整根跳过,不拿别的字段凑。"""
    out: List[float] = []
    for bar in bars or []:
        value = bar.get(key)
        if isinstance(value, (int, float)) and value > 0:
            out.append(float(value))
    return out


def trend_levels(
    bars: Optional[Sequence[Dict[str, Any]]],
    spot: float,
    ma_periods: Sequence[int] = DEFAULT_MA_PERIODS,
    extreme_window: int = EXTREME_WINDOW,
) -> List[AlertLevel]:
    """日线均线 + 52 周高低点 → 价位。bars 为按日期升序的前复权日 K。

    这些位置和期权墙一样"被全市场盯着":跌到年线、创 52 周新低,都是大量
    资金共用的历史低位标尺。均线用收盘价简单平均;高低点用最高/最低价——
    衡量"历史上到过哪儿"要用真实摸到过的价格,不是收盘价。

    数据不够就少给:均线凑不满周期的不算(拿 30 根算"60日均线"是撒谎);
    历史不足 52 周时高低点降级为"上市以来",标签也跟着改——降级可以,
    但不能假装是 52 周。
    """
    if not bars or spot is None or spot <= 0:
        return []
    out: List[AlertLevel] = []

    closes = _bar_values(bars, "close")
    for period in ma_periods:
        if len(closes) < period:
            continue
        ma = round(sum(closes[-period:]) / period, 4)
        if ma <= 0:
            continue
        out.append(AlertLevel(ma, "%d日均线" % period, "ma%d" % period,
                              _trend_kind(ma, spot), _PRIORITY_TREND))

    lows = _bar_values(bars, "low")
    highs = _bar_values(bars, "high")
    if len(lows) >= MIN_EXTREME_BARS and len(highs) >= MIN_EXTREME_BARS:
        full = len(lows) >= extreme_window and len(highs) >= extreme_window
        low = round(min(lows[-extreme_window:]), 4)
        high = round(max(highs[-extreme_window:]), 4)
        out.append(AlertLevel(low, "52周低点" if full else "历史低点", "low_52w",
                              _trend_kind(low, spot), _PRIORITY_TREND))
        out.append(AlertLevel(high, "52周高点" if full else "历史高点", "high_52w",
                              _trend_kind(high, spot), _PRIORITY_TREND))
    return out


def trend_snapshot(
    bars: Optional[Sequence[Dict[str, Any]]],
    spot: float,
    ma_periods: Sequence[int] = DEFAULT_MA_PERIODS,
    extreme_window: int = EXTREME_WINDOW,
) -> Optional[Dict[str, Any]]:
    """给界面看的趋势摘要:各均线值、52 周高低点、现价在一年区间里的位置。

    range_pos_pct 是"历史低位"的量化读数:0 = 贴着 52 周低点,100 = 贴着高点。
    """
    levels = trend_levels(bars, spot, ma_periods, extreme_window)
    if not levels:
        return None
    out: Dict[str, Any] = {level.source: level.price for level in levels}
    low, high = out.get("low_52w"), out.get("high_52w")
    if low is not None and high is not None and high > low:
        out["range_pos_pct"] = round((spot - low) / (high - low) * 100.0, 1)
    out["bars"] = len(_bar_values(bars, "close"))
    return out


def build_levels(
    spot: float,
    wall: Optional[Dict[str, Any]] = None,
    step: float = DEFAULT_STEP,
    merge_pct: float = DEFAULT_MERGE_PCT,
    max_levels: int = MAX_LEVELS,
    history: Optional[Sequence[Dict[str, Any]]] = None,
) -> List[AlertLevel]:
    """期权墙价位 + 趋势位 + 整数关口,去重合并后按价格升序返回。

    合并很重要:墙正好落在整数关口上是常事(它们本来就互相吸引),
    不合并的话同一个价位会报两次——而"报两次"和"报错"在使用上是一回事。
    """
    if spot <= 0:
        raise AlertError("缺少现价,无法生成价位。")

    levels = _wall_levels(wall)
    levels += trend_levels(history, spot)
    levels += [
        AlertLevel(price, "整数关口 %g" % price, "round", "pivot", _PRIORITY_ROUND)
        for price in round_levels(spot, step)
    ]

    tol = abs(spot) * merge_pct / 100.0
    merged: List[AlertLevel] = []
    for level in sorted(levels, key=lambda l: l.price):
        if merged and abs(level.price - merged[-1].price) <= tol:
            previous = merged[-1]
            # 以更可信的那个为准:墙压过整数关口,0DTE 下成交墙压过持仓墙。
            # 标签两个都留着,让人看得见这个位是被几件事同时指着的。
            winner, loser = (
                (previous, level) if previous.priority >= level.priority
                else (level, previous)
            )
            merged[-1] = AlertLevel(
                winner.price,
                "%s + %s" % (winner.label, loser.label),
                winner.source,
                winner.kind if winner.kind != "pivot" else loser.kind,
                winner.priority,
            )
            continue
        merged.append(level)

    # 超出上限时,留离现价最近的那些——远处的位轮不到今天用
    if len(merged) > max_levels:
        merged = sorted(merged, key=lambda l: abs(l.price - spot))[:max_levels]
    return sorted(merged, key=lambda l: l.price)


# ---------------------------------------------------------------- 状态机
def _band(price: float, band_pct: float) -> float:
    return abs(price) * band_pct / 100.0


def evaluate(
    levels: Sequence[AlertLevel],
    states: Dict[str, LevelState],
    prev_price: Optional[float],
    price: float,
    now_ts: float,
    band_pct: float = DEFAULT_BAND_PCT,
    cooldown: float = DEFAULT_COOLDOWN,
) -> Tuple[List[Dict[str, Any]], Dict[str, LevelState]]:
    """价格从 prev_price 走到 price,哪些价位该报警。

    报警条件是**穿过**,不是"接近":两次取价之间跨过了这个价位才算。
    穿过之后立刻落防,直到价格离开 band 之外才重新上膛;再加一层冷却时间,
    防的是价格恰好在 band 边缘反复进出的情况。

    第一次调用(prev_price 为 None)只登记价格,不报警——否则一打开页面
    就会把现价附近的价位全部报一遍。
    """
    events: List[Dict[str, Any]] = []
    out = dict(states)

    for level in levels:
        key = level_key(level)
        state = out.get(key) or LevelState()
        band = _band(level.price, band_pct)

        # 先看能不能重新上膛:离开得够远,而且过了冷却
        if not state.armed and abs(price - level.price) > band:
            cooled = (
                state.last_fired_at is None or now_ts - state.last_fired_at >= cooldown
            )
            if cooled:
                state = LevelState(armed=True, last_fired_at=state.last_fired_at)

        if prev_price is not None and state.armed:
            low, high = min(prev_price, price), max(prev_price, price)
            if low <= level.price <= high:
                direction = "up" if price >= prev_price else "down"
                events.append(
                    {
                        "price": level.price,
                        "label": level.label,
                        "source": level.source,
                        "kind": level.kind,
                        "direction": direction,
                        "from": round(prev_price, 4),
                        "to": round(price, 4),
                        "at": now_ts,
                        "text": "%s %s %s(现价 %s)"
                        % (level.label, "上穿" if direction == "up" else "下破",
                           _fmt(level.price), _fmt(price)),
                    }
                )
                state = LevelState(armed=False, last_fired_at=now_ts)

        out[key] = state

    # 价位重算之后老状态会变成孤儿,留着只会越攒越多
    valid = {level_key(l) for l in levels}
    return events, {k: v for k, v in out.items() if k in valid}


def level_key(level: AlertLevel) -> str:
    """状态字典的键。只用价格:墙的标签会随行情变,价格才是这个警报的身份。"""
    return "%.4f" % level.price


def _fmt(value: float) -> str:
    return ("%.4f" % value).rstrip("0").rstrip(".")


def describe(levels: Sequence[AlertLevel], spot: float) -> List[str]:
    """给人看的一句话清单,界面和通知都用它。"""
    out: List[str] = []
    for level in levels:
        gap = (level.price / spot - 1.0) * 100.0 if spot else 0.0
        out.append("%s %s(%+.2f%%)" % (_fmt(level.price), level.label, gap))
    return out

File: engine-ts/src/alerts.ts
/** 价位警告(对应 Python alerts.py)—— 纯计算 + 状态机。
 *
 * 最难的不是"设警报",是"不要刷屏":每个价位独立状态机,
 * 穿过 → 报一次 → 落防 → 离开 band 之外且过冷却 → 重新上膛。
 */
import type { LevelKind, LevelState, TrendSnapshot, WatchEvent, WatchLevel } from "./contract/alerts.js";
import { fmtF, fmtSF, pyRound } from "./py.js";

// 价位的报警状态定义在 contract/alerts.ts(盯单上存的就是它);这里转出,老的 import 不用改。
export type { LevelState } from "./contract/alerts.js";

export const DEFAULT_STEP = 5.0;
export const DEFAULT_MERGE_PCT = 0.15;
export const DEFAULT_BAND_PCT = 0.3;
export const DEFAULT_COOLDOWN = 300.0;
export const MAX_LEVELS = 24;

export const DEFAULT_MA_PERIODS: readonly number[] = [20, 60, 120, 200];   // 与 Python DEFAULT_MA_PERIODS 同步
export const EXTREME_WINDOW = 250; // 52 周 ≈ 250 个交易日
export const MIN_EXTREME_BARS = 20; // 历史太短时高低点没意义,干脆不给

export class AlertError extends Error {}

/** 计算过程里的价位:盯单上存的那四样(WatchLevel)+ 合并时用的可信度,priority 不上线、不落库。 */
export interface AlertLevel extends WatchLevel {
  priority: number;
}

export function levelDict(l: AlertLevel): WatchLevel {
  return { price: l.price, label: l.label, source: l.source, kind: l.kind };
}

// ---------------------------------------------------------------- 价位生成
/** 现价上下最近的整数关口。现价正好在关口上时取上下各一档。 */
export function roundLevels(spot: number, step = DEFAULT_STEP): number[] {
  if (spot <= 0 || step <= 0) return [];
  const below = Math.floor(spot / step) * step;
  const above = Math.ceil(spot / step) * step;
  if (Math.abs(above - below) < 1e-9) {
    return [pyRound(below - step, 6), pyRound(above + step, 6)];
  }
  return [pyRound(below, 6), pyRound(above, 6)];
}

const PRIORITY_ROUND = 0;
const PRIORITY_PIVOT = 1;
const PRIORITY_WEAK_WALL = 2;
// 趋势位(均线/52周高低点)与弱墙同级:全市场看得见、但不带当天持仓依据
const PRIORITY_TREND = 2;
const PRIORITY_STRONG_WALL = 3;

function wallLevels(wall: Record<string, any> | null | undefined): AlertLevel[] {
  if (!wall) return [];
  const out: AlertLevel[] = [];
  const zeroDay = (wall["days_to_expiry"] ?? 99) <= 1.0;
  const volRank = zeroDay ? PRIORITY_STRONG_WALL : PRIORITY_WEAK_WALL;
  const oiRank = zeroDay ? PRIORITY_WEAK_WALL : PRIORITY_STRONG_WALL;

  const add = (key: string, label: string, rank: number): void => {
    const entry = wall[key];
    if (entry && entry["strike"]) {
      const kind: LevelKind = key.includes("call") ? "resistance" : "support";
      out.push({ price: Number(entry["strike"]), label, source: key, kind, priority: rank });
    }
  };

  add("call_vol_wall", "上方成交墙", volRank);
  add("put_vol_wall", "下方成交墙", volRank);
  add("call_wall", "上方持仓墙", oiRank);
  add("put_wall", "下方持仓墙", oiRank);

  if (wall["max_pain"]) {
    out.push({
      price: Number(wall["max_pain"]["strike"]), label: "最大痛点",
      source: "max_pain", kind: "pivot", priority: PRIORITY_PIVOT,
    });
  }
  if (wall["gamma_flip"] !== null && wall["gamma_flip"] !== undefined) {
    out.push({
      price: Number(wall["gamma_flip"]), label: "Gamma 翻转位",
      source: "gamma_flip", kind: "pivot", priority: PRIORITY_PIVOT,
    });
  }
  return out;
}

function trendKind(price: number, spot: number): LevelKind {
  if (price < spot) return "support";
  if (price > spot) return "resistance";
  return "pivot";
}

/** 按顺序取一列有效价格。缺了或非正数的整根跳过,不拿别的字段凑。 */
function barValues(bars: Array<Record<string, any>> | null | undefined, key: string): number[] {
  const out: number[] = [];
  for (const bar of bars ?? []) {
    const value = bar[key];
    if (typeof value === "number" && Number.isFinite(value) && value > 0) out.push(value);
  }
  return out;
}

/** 日线均线 + 52 周高低点 → 价位(对应 Python trend_levels,含标签降级规则)。 */
export function trendLevels(
  bars: Array<Record<string, any>> | null | undefined,
  spot: number,
  maPeriods: readonly number[] = DEFAULT_MA_PERIODS,
  extremeWindow = EXTREME_WINDOW,
): AlertLevel[] {
  if (!bars || !bars.length || !spot || spot <= 0) return [];
  const out: AlertLevel[] = [];

  const closes = barValues(bars, "close");
  for (const period of maPeriods) {
    if (closes.length < period) continue;
    const window = closes.slice(-period);
    const ma = pyRound(window.reduce((a, b) => a + b, 0) / period, 4);
    if (ma <= 0) continue;
    out.push({
      price: ma, label: `${period}日均线`, source: `ma${period}`,
      kind: trendKind(ma, spot), priority: PRIORITY_TREND,
    });
  }

  const lows = barValues(bars, "low");
  const highs = barValues(bars, "high");
  if (lows.length >= MIN_EXTREME_BARS && highs.length >= MIN_EXTREME_BARS) {
    const full = lows.length >= extremeWindow && highs.length >= extremeWindow;
    const low = pyRound(Math.min(...lows.slice(-extremeWindow)), 4);
    const high = pyRound(Math.max(...highs.slice(-extremeWindow)), 4);
    out.push({
      price: low, label: full ? "52周低点" : "历史低点", source: "low_52w",
      kind: trendKind(low, spot), priority: PRIORITY_TREND,
    });
    out.push({
      price: high, label: full ? "52周高点" : "历史高点", source: "high_52w",
      kind: trendKind(high, spot), priority: PRIORITY_TREND,
    });
  }
  return out;
}

/** 趋势摘要:均线值、52 周高低点、现价在一年区间里的位置(0 = 贴着低点)。 */
export function trendSnapshot(
  bars: Array<Record<string, any>> | null | undefined,
  spot: number,
  maPeriods: readonly number[] = DEFAULT_MA_PERIODS,
  extremeWindow = EXTREME_WINDOW,
): TrendSnapshot | null {
  const levels = trendLevels(bars, spot, maPeriods, extremeWindow);
  if (!levels.length) return null;
  const out: TrendSnapshot = {};
  for (const level of levels) out[level.source] = level.price;
  const low = out["low_52w"];
  const high = out["high_52w"];
  if (low !== undefined && high !== undefined && high > low) {
    out["range_pos_pct"] = pyRound(((spot - low) / (high - low)) * 100.0, 1);
  }
  out["bars"] = barValues(bars, "close").length;
  return out;
}

/** 期权墙价位 + 趋势位 + 整数关口,去重合并后按价格升序返回。 */
export function buildLevels(
  spot: number,
  wall: Record<string, any> | null = null,
  step = DEFAULT_STEP,
  mergePct = DEFAULT_MERGE_PCT,
  maxLevels = MAX_LEVELS,
  history: Array<Record<string, any>> | null = null,
): AlertLevel[] {
  if (spot <= 0) throw new AlertError("缺少现价,无法生成价位。");

  const levels = wallLevels(wall);
  levels.push(...trendLevels(history, spot));
  for (const price of roundLevels(spot, step)) {
    levels.push({
      price, label: `整数关口 ${pyG(price)}`, source: "round", kind: "pivot",
      priority: PRIORITY_ROUND,
    });
  }

  const tol = (Math.abs(spot) * mergePct) / 100.0;
  const merged: AlertLevel[] = [];
  for (const level of [...levels].sort((a, b) => a.price - b.price)) {
    if (merged.length && Math.abs(level.price - merged[merged.length - 1]!.price) <= tol) {
      const previous = merged[merged.length - 1]!;
      // 以更可信的那个为准,标签两个都留着
      const [winner, loser] =
        previous.priority >= level.priority ? [previous, level] : [level, previous];
      merged[merged.length - 1] = {
        price: winner.price,
        label: `${winner.label} + ${loser.label}`,
        source: winner.source,
        kind: winner.kind !== "pivot" ? winner.kind : loser.kind,
        priority: winner.priority,
      };
      continue;
    }
    merged.push(level);
  }

  let out = merged;
  if (out.length > maxLevels) {
    out = [...out]
      .sort((a, b) => Math.abs(a.price - spot) - Math.abs(b.price - spot))
      .slice(0, maxLevels);
  }
  return [...out].sort((a, b) => a.price - b.price);
}

// ---------------------------------------------------------------- 状态机
function band(price: number, bandPct: number): number {
  return (Math.abs(price) * bandPct) / 100.0;
}

/** 一次穿越,还不知道是哪只股的(symbol 由调用方按盯单补上,补完就是 WatchEvent)。 */
export type CrossingEvent = Omit<WatchEvent, "symbol">;

/** 价格从 prevPrice 走到 price,哪些价位该报警。第一次调用只登记不报警。 */
export function evaluate(
  levels: AlertLevel[],
  states: Record<string, LevelState>,
  prevPrice: number | null,
  price: number,
  nowTs: number,
  bandPct = DEFAULT_BAND_PCT,
  cooldown = DEFAULT_COOLDOWN,
): [CrossingEvent[], Record<string, LevelState>] {
  const events: CrossingEvent[] = [];
  const out: Record<string, LevelState> = { ...states };

  for (const level of levels) {
    const key = levelKey(level);
    let state = out[key] ?? { armed: true, last_fired_at: null };
    const bd = band(level.price, bandPct);

    // 先看能不能重新上膛:离开得够远,而且过了冷却
    if (!state.armed && Math.abs(price - level.price) > bd) {
      const cooled = state.last_fired_at === null || nowTs - state.last_fired_at >= cooldown;
      if (cooled) state = { armed: true, last_fired_at: state.last_fired_at };
    }

    if (prevPrice !== null && state.armed) {
      const low = Math.min(prevPrice, price);
      const high = Math.max(prevPrice, price);
      if (low <= level.price && level.price <= high) {
        const direction: "up" | "down" = price >= prevPrice ? "up" : "down";
        events.push({
          price: level.price,
          label: level.label,
          source: level.source,
          kind: level.kind,
          direction,
          from: pyRound(prevPrice, 4),
          to: pyRound(price, 4),
          at: nowTs,
          text:
            `${level.label} ${direction === "up" ? "上穿" : "下破"} ${fmtStrip(level.price)}` +
            `(现价 ${fmtStrip(price)})`,
        });
        state = { armed: false, last_fired_at: nowTs };
      }
    }
    out[key] = state;
  }

  // 价位重算之后老状态会变成孤儿,清掉
  const valid = new Set(levels.map(levelKey));
  const cleaned: Record<string, LevelState> = {};
  for (const [k, v] of Object.entries(out)) if (valid.has(k)) cleaned[k] = v;
  return [events, cleaned];
}

/** 状态字典的键。只用价格:价格才是这个警报的身份。 */
export function levelKey(level: AlertLevel): string {
  return fmtF(level.price, 4);
}

function fmtStrip(value: number): string {
  return fmtF(value, 4).replace(/0+$/, "").replace(/\.$/, "");
}

/** 给人看的一句话清单。 */
export function describe(levels: AlertLevel[], spot: number): string[] {
  return levels.map((level) => {
    const gap = spot ? (level.price / spot - 1.0) * 100.0 : 0.0;
    return `${fmtStrip(level.price)} ${level.label}(${fmtSF(gap, 2)}%)`;
  });
}

import { pyG } from "./py.js";
